### ISSUE TITLE
[cuda] lineinfo for device function

### ISSUE THREAD

**Issue by sklam** [numba/numba#5311]
Related gitter discussion starting at [here](https://gitter.im/numba/numba?at=5e4ff1bd8b2d4664ef17be48) and [here](https://gitter.im/numba/numba?at=5e53d9291be0ff01d5a69887).

Reproducer:

```python
# filename: debugme.py
import math
import numpy as np
import numba
from numba import cuda


@cuda.jit(device=True, debug=True)
def expensive(x):
    return math.pow(x, 1.001)


@cuda.jit(device=True, debug=True)
def cheap(x):
    return math.sqrt(x)


@cuda.jit(device=True, debug=True)
def calc(x):
    s1 = expensive(x)
    q = cheap(s1)
    return q + 1


@cuda.jit(device=True, debug=True)
def threadid():
    return cuda.blockDim.x * cuda.blockIdx.x + cuda.threadIdx.x


@cuda.jit(argtypes=[numba.int32[:]], debug=True)
def kernel(arr):
    for i in range(1000):
        where = threadid()
        value = arr[where]
        arr[where] = calc(value)


def main():
    print("SUP")
    z = np.array(range(0, 5000))
    print(z)
    kernel[5000 // 128 + 1, 128](z)
    print(z)

    print(kernel.inspect_llvm())

if __name__ == '__main__':
    main()
```

Run the above with:

```bash
$ nvprof -a instruction_execution --export-profile profile_output python debugme.py
```

On my OSX 10.13 with CUDA 9.2, we can see some line-info in the `profile_output` using NVVP:

<img width="521" alt="separate_devfns" src="https://user-images.githubusercontent.com/1929845/75255060-faac3980-57a6-11ea-8b22-448fea36f026.png">

The "inline-closure" version (this style: https://gist.github.com/c200chromebook/d131322c918b142f14f40c9b3583cc6d) looks like:
<img width="532" alt="inlined_closure" src="https://user-images.githubusercontent.com/1929845/75255017-e7996980-57a6-11ea-8873-7a0f312c860c.png">

@c200chromebook, can you tell us what you see?

**Comment by c200chromebook** [comment]
CUDA 10.1, numba 0.48.0, numpy 1.18.1, llvmlite 0.31.0 - you don't even get near the profiling.

![image](https://user-images.githubusercontent.com/9907030/75358072-4da5ef80-5880-11ea-8e52-db03e415cc86.png)

**Comment by c200chromebook** [comment]
Using the goofy inline-closure format:
![image](https://user-images.githubusercontent.com/9907030/75359175-07519000-5882-11ea-8e50-dc3edf59e8e4.png)

**Comment by sklam** [comment]
hmm... i think this might be a problem in libNVVM in CUDA 10.1. @gmarkall, do you know if there are known problems regarding debug-symbols in newer CUDA.

**Comment by c200chromebook** [comment]
Yeah, you'll excuse me, I did very minorly goof - 10.2. Forgot we had upgraded.

**Comment by gmarkall** [comment]
@sklam If you're referring to the appearance of the message "Debugging support cannot be enabled when the number of debug compile units is more than 1", it's a message I've bumped into before but hadn't yet got around to looking into... I didn't realise it ever did work with 9.2 though.

**Comment by stuartarchibald** [comment]
I've also seen that message a fair few times.

**Comment by sklam** [comment]
@stuartarchibald, on what hardwares and CUDA versions?

**Comment by gmarkall** [comment]
I've seen it with CUDA 10.0 and a V100.

**Comment by stuartarchibald** [comment]
> @stuartarchibald, on what hardwares and CUDA versions?

Stretching memory, but CUDA 10.2 and a GTX970.

**Comment by gmarkall** [comment]
> hmm... i think this might be a problem in libNVVM in CUDA 10.1. @gmarkall, do you know if there are known problems regarding debug-symbols in newer CUDA.

I just noticed that it's documented that only a single module is supported: https://docs.nvidia.com/cuda/nvvm-ir-spec/index.html#source-level-debugging-support:

> Source level debugging is supported only for a single debug compile unit. If there are multiple input NVVM IR modules, at most one module may have a single debug compile unit.

**Comment by sklam** [comment]
Hm... we can possibly prelink the module in llvm before sending it over to nvvm.

**Comment by c200chromebook** [comment]
Is this one a fundamental limitation or should the prelink workaround fix?

**Comment by sklam** [comment]
@gmarkall, the doc change first happened in 10.0 (https://docs.nvidia.com/cuda/archive/10.0/nvvm-ir-spec/index.html#source-level-debugging-support). In 9.2 (https://docs.nvidia.com/cuda/archive/9.2/nvvm-ir-spec/index.html#source-level-debugging-support), there were no restriction on "a single debug compile unit" and it works. 

Why is this restriction needed? I would think that one can use LLVM to link the modules together and the debug symbol will merge fine. If it is not that simple, I don't know what Numba can do to address this problem.

@c200chromebook is reporting (at https://gitter.im/numba/numba?at=5f35a6314bb58d3181951aea) that this affects a single kernel case in Numba. Which is odd because there should only be one LLVM module. Unless NVVM is counting the libdevice that we are linking in.

**Comment by c200chromebook** [comment]
See also the below, which has two items to compile, thus causing a fail.
```
from numba import cuda
import numba as nb

@cuda.jit(nb.void(nb.int32[:], nb.int32[:]), debug=True)
def whatever(inp, outp):
    outp[0] = 1 if inp[0] in (2, 3) else 3


whatever[1, 1](cuda.device_array(1), cuda.device_array(1))
```
yields
```
Error: Debugging support cannot be enabled when number of debug compile units is more than 1
NVVM_ERROR_COMPILATION
```

**Comment by gmarkall** [comment]
On master with CUDA 11.2 (NVVM 7.0) line info seems to be available again with `debug=True`:

![image](https://user-images.githubusercontent.com/535640/111703332-68edef00-8835-11eb-9dd2-5382698941c1.png)

**Comment by gmarkall** [comment]
> See also the below, which has two items to compile, thus causing a fail.
> 
> ```
> from numba import cuda
> import numba as nb
> 
> @cuda.jit(nb.void(nb.int32[:], nb.int32[:]), debug=True)
> def whatever(inp, outp):
>     outp[0] = 1 if inp[0] in (2, 3) else 3
> 
> 
> whatever[1, 1](cuda.device_array(1), cuda.device_array(1))
> ```
> 
> yields
> 
> ```
> Error: Debugging support cannot be enabled when number of debug compile units is more than 1
> NVVM_ERROR_COMPILATION
> ```

This reproduces for me with CUDA 10.1, but the issue appears to be resolved with CUDA 11.2.

**Comment by gmarkall** [comment]
To summarise where I think we are:

- Our debuginfo generation is better for NVVM 7.0 than it is for NVVM 3.4, due to using the exact same mechanism as the CPU target.
- For a presently-unknown reason, the "Error: Debugging support cannot be enabled when number of debug compile units is more than 1" is not triggered with NVVM 7.0, though this is still unsupported: https://docs.nvidia.com/cuda/archive/11.2.1/nvvm-ir-spec/index.html#source-level-debugging-support
- See also the PR to add `lineinfo` generation only: https://github.com/numba/numba/pull/6802

@sklam asked earlier:

> Why is this restriction needed? I would think that one can use LLVM to link the modules together and the debug symbol will merge fine. If it is not that simple, I don't know what Numba can do to address this problem.

I don't know the reason for the restriction. I believe that we can conform with this restriction in Numba by changing 

https://github.com/numba/numba/blob/81f708263be0ba6e803bd54633e8c278d93a5e78/numba/cuda/codegen.py#L114-L115

so that instead of passing a list of the IR for all modules into `llvm_to_ptx`, we call it once for each module's IR, then link the resulting PTX for each module one-by-one when linking the cubin. So here:

https://github.com/numba/numba/blob/81f708263be0ba6e803bd54633e8c278d93a5e78/numba/cuda/codegen.py#L139

we'd have multiple PTXs to add.

I hadn't done this by default so far because it seems to result in less inlining and less efficient code in the non-debug case (I did try it, but couldn't get the linker to inline the PTX as eagerly as NVVM seems to inline IR for device functions). For debug, this isn't a problem though, as we're not going to be expecting much / any inlining.

We have this hack in the code:

https://github.com/numba/numba/blob/81f708263be0ba6e803bd54633e8c278d93a5e78/numba/cuda/cudadrv/nvvm.py#L708-L728

which is intended to workaround the fact we are passing in multiple modules when `debug=True`. If the above changes to pass one module to NVVM at a time are implemented, I believe that this workaround could be removed (and maybe debug info in CUDA would work a bit better than it presently does).

**Comment by gmarkall** [comment]
> instead of passing a list of the IR for all modules into llvm_to_ptx, we call it once for each module's IR, then link the resulting PTX for each module one-by-one when linking the cubin.

I've tried implementing this strategy and removing the debug_pubnames patching in: https://github.com/gmarkall/numba/commit/ce840bc90e8d1ed6831ae797e50f0d0d06bde091

It fixes the failure to compile with NVVM 3.4. Things seem to work as well as they did before with NVVM 7.0. Stepping line-by-line seems like it might be more stable, but I'm not sure if that's just my perception - I need to do some more careful experiments.

**Comment by gmarkall** [comment]
PR #6841 includes a fix for the message "Error: Debugging support cannot be enabled when number of debug compile units is more than 1" using the strategy described above - since lineinfo is working with NVVM 7.0, and the error message is resolved with NVVM 3.4, I will consider this issue resolved once that PR is merged.
